program/src/processor.rs

```diff
@@ -215,12 +209,51 @@ fn process_cancel_proposal(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
 
     check_proposal_exists(program_id, proposal_info)?;
 
+    let mut proposal_data = proposal_info.try_borrow_mut_data()?;
+    let proposal_state = bytemuck::try_from_bytes_mut::<Proposal>(&mut proposal_data)
+        .map_err(|_| ProgramError::InvalidAccountData)?;
+
+    // Ensure the stake authority is the proposal author.
+    proposal_state.check_author(stake_authority_info.key)?;
+
+    // Ensure the proposal is in draft or voting stage.
+    if proposal_state.status != ProposalStatus::Draft
+        && proposal_state.status != ProposalStatus::Voting
+    {
+        return Err(PaladinGovernanceError::ProposalIsImmutable.into());
+    }
```

Comment:
Yeah exactly. I always think state machines and exhaustive enum checks should go hand in hand to be totally sure that no cases are missed, ie. making it possible to cancel a rejected proposal or something

Reply:
I've added a commit for this, but only for this processor. The others explicitly only allow one variant (ie. `Voting`), so this one I felt like is the only one that makes sense to be explicitly exhaustive, in case someone wants to allow cancelling proposals in other stages.

Reply:
makes sense to me, thanks!